sstables/sstables_manager.cc

```diff
@@ -149,8 +149,40 @@ sstable_writer_config sstables_manager::configure_writer(sstring origin) const {
     return cfg;
 }
 
-void sstables_manager::update_total_memory_used_by_components(const sstable* sst) {
+void sstables_manager::update_total_memory_used_by_components_and_maybe_reclaim(const sstable* sst) {
     _total_memory_used_by_components += sst->total_components_memory_size();
+
+    if (_total_memory_used_by_components <= _max_allowed_memory_for_components) {
```

Comment:
We need a symmetric hook on the sstable delete path: if the total memory goes back below the limit, we can reload some components in the background.
Maybe this should be a follow-up, because this will require adding synchronization between reclaim and reload and we want to keep this first PR simple so we can backport it.

Reply:
Ack. We would have to track all the discarded components and then load them when a SSTable is deactivated and the memory becomes available. I agree its better to do it as a follow-up to this.